source/expectations/ut_expectation_base.tpb

```diff
@@ -0,0 +1,44 @@
+create or replace type body ut_expectation_base as
```

Comment:
I wonder if both `ut_expectation_base` as well as `ut_matcher_base` are required.
Normally, you cannot have a circular dependency and to overcome that one of type needs to have a `_base` and the other must depend on `_base`.

I would need to have a look in depth to see if the solution would works with single base type.
It would also be good to figure out pros/cons of having `ut_machter_base` vs `ut_expectation_base` to choose the one that fits better.

We might consider the `_base` type to be an interface. Something Oracle types don't have really.
It would definitely be an interface if it would be marked as non instantiable.

Reply:
Unfortunately the matcher have expectation as attribute and expectation is using matcher to call to_ method.